File: src/dotnet/FilterIcs/Manifest.cs
using System.Text.Json;

namespace FilterIcs;

/// <summary>Feed manifest (public/.feeds.json): basename -> output name.</summary>
public static class Manifest
{
    public static Dictionary<string, string> Read(string path)
    {
        if (!File.Exists(path))
            return new Dictionary<string, string>();
        try
        {
            string text = File.ReadAllText(path);
            var map = JsonSerializer.Deserialize<Dictionary<string, string>>(text);
            return map ?? new Dictionary<string, string>();
        }
        catch (Exception ex) when (ex is JsonException or IOException)
        {
            return new Dictionary<string, string>();
        }
    }

    public static void Write(string path, IDictionary<string, string> map)
    {
        var sorted = new SortedDictionary<string, string>(new Dictionary<string, string>(map), StringComparer.Ordinal);
        string json = JsonSerializer.Serialize(sorted, new JsonSerializerOptions { WriteIndented = true }) + "\n";
        Render.AtomicWrite(path, json);
    }
}

File: src/python/filter_ics/manifest.py
"""Feed manifest (``public/.feeds.json``): basename -> output name.

Used to distinguish intentional removals (source file gone) from transient per-feed
errors. See contracts/publishing.md.
"""

from __future__ import annotations

import json
from pathlib import Path

from .render import atomic_write


def read_manifest(path: str | Path) -> dict[str, str]:
    """Read the manifest; return {} if missing or unreadable/invalid."""
    p = Path(path)
    if not p.exists():
        return {}
    try:
        data = json.loads(p.read_text(encoding="utf-8"))
    except (json.JSONDecodeError, OSError):
        return {}
    if not isinstance(data, dict):
        return {}
    return {str(k): str(v) for k, v in data.items()}


def write_manifest(path: str | Path, mapping: dict[str, str]) -> None:
    """Write the manifest atomically (sorted keys, UTF-8)."""
    payload = json.dumps(mapping, indent=2, ensure_ascii=False, sort_keys=True) + "\n"
    atomic_write(path, payload.encode("utf-8"))
